postgis/lwgeom_functions_basic.c

```diff
@@ -2399,6 +2400,124 @@ Datum LWGEOM_azimuth(PG_FUNCTION_ARGS)
 	PG_RETURN_FLOAT8(result);
 }
 
+/**
+ * Compute the angle defined by 3 points or the angle between 2 vectors
+ * defined by 4 points
+ * given Point geometries.
+ * @return NULL on exception (same point).
+ * 		Return radians otherwise (always positive).
+ */
+PG_FUNCTION_INFO_V1(LWGEOM_angle);
+Datum LWGEOM_angle(PG_FUNCTION_ARGS)
+{
+	GSERIALIZED *geom;
+	LWGEOM *geom_unser;
+	LWPOINT *lwpoint;
+	POINT2D points[4]; 
+	double az1,az2 ;
+	double result;
+	int srids[4];
+	int i = 0 ; 
+	int err_code = 0; 
+	int n_args = PG_NARGS(); 
+
+	/* no deserialize, checking for common error first*/
+	for(i=0; i<n_args; i++)
+	{
+		geom = PG_GETARG_GSERIALIZED_P(i); 
+		if (gserialized_is_empty(geom) )
+		{/* empty geom */
+			if (i==3)
+			{
+				n_args = 3 ;
+			}
+			else
+			{
+				err_code = 1 ;
+				break ;
+			}
+		} else
+		{
+			if(gserialized_get_type(geom) != POINTTYPE)
+			{/* geom type */
+				err_code = 2 ;
+				break;
+			}
+			else
+			{
+				srids[i] = gserialized_get_srid(geom) ; 
+				if(srids[0] != srids[i])
+				{/* error on srid*/
+					err_code = 3 ;
+					break;
+				}
+			}
+		}
+		PG_FREE_IF_COPY(geom, i);
+	}
+	if (err_code >0)
+		switch (err_code){
+			default: /*always executed*/
+			PG_FREE_IF_COPY(geom, i);
+			
+			case 1:
+			lwpgerror("Empty geometry"); 
+			PG_RETURN_NULL() ;
+			break;
+			
+			case 2:
+			lwpgerror("Argument must be POINT geometries");
+			PG_RETURN_NULL();
+			break;
+			
+			case 3:
+			lwpgerror("Operation on mixed SRID geometries");
+			PG_RETURN_NULL();
+			break;
+		}
+	/* extract points */
+	for(i=0; i<n_args; i++)
+	{
+		geom = PG_GETARG_GSERIALIZED_P(i);
+		geom_unser = lwgeom_from_gserialized(geom) ; 
+		lwpoint = lwgeom_as_lwpoint(geom_unser); 
+		if (!lwpoint)
+		{ 
+			lwpgerror("Error unserializing geometry"); 
+			PG_RETURN_NULL() ;
```

Comment:
@strk Ok as you suggested I stored the serialized pointers to keep getting
those twice.
Now `PG_FREE_IF_COPY` is always executed if exceptions or in regular
operations.

2016-02-26 14:47 GMT+01:00 Sandro Santilli <notifications@github.com>:

> In postgis/lwgeom_functions_basic.c
> <https://github.com/postgis/postgis/pull/97#discussion_r54244102>:
>
> > +			
> > +			case 3:
> > +			lwpgerror("Operation on mixed SRID geometries");
> > +			PG_RETURN_NULL();
> > +			break;
> > +		}
> > +	/* extract points */
> > +	for(i=0; i<n_args; i++)
> > +	{
> > +		geom = PG_GETARG_GSERIALIZED_P(i);
> > +		geom_unser = lwgeom_from_gserialized(geom) ;
> > +		lwpoint = lwgeom_as_lwpoint(geom_unser);
> > +		if (!lwpoint)
> > +		{
> > +			lwpgerror("Error unserializing geometry");
> > +			PG_RETURN_NULL() ;
>
> Also this misses a PG_FREE_IF_COPY...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/postgis/postgis/pull/97/files#r54244102>.
>